cpp/log/cluster_state_controller.h

```diff
@@ -0,0 +1,92 @@
+#ifndef CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_H_
+#define CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_H_
+
+#include <condition_variable>
+#include <functional>
+#include <map>
+#include <string>
+
+#include "log/etcd_consistent_store.h"
+#include "proto/ct.pb.h"
+#include "util/executor.h"
+#include "util/masterelection.h"
+
+namespace cert_trans {
+
+
+// A class which updates & maintains the states of the individual cluster
+// member nodes, and uses this information to determine the overall serving
+// state of the cluster.
+//
+// In particular, this class calculates the optimal STH for the cluster to
+// serve at any given time.
+template <class Logged>
+class ClusterStateController {
+ public:
+  ClusterStateController(util::Executor* executor,
+                         ConsistentStore<Logged>* store,
+                         const MasterElection* election,
+                         const int min_serving_nodes,
+                         const double min_serving_fraction);
+
+  ~ClusterStateController();
+
+  // Updates *this* node's ClusterNodeState to reflect the new STH available.
+  void NewTreeHead(const ct::SignedTreeHead& sth);
+
+  // Updates *this* node's ClusterNodeState to reflect the new tree data
+  // available.
+  void ContiguousTreeSizeUpdated(const int64_t new_contiguous_tree_size);
+
+  // Gets the current (if any) calculated serving STH for the cluster.
+  // If there is such an STH then return true and |sth| is populated, returns
+  // false otherwise.
+  //
+  // Note that this simply returns this node's intepretation of the optimum
+  // serving STH, the current master/contents of the servingSTH file may
+  // differ.
+  //
+  // Really only intended for testing.
+  bool GetCalculatedServingSTH(ct::SignedTreeHead* sth) const;
```

Comment:
Hmm, that's of marginal value in this particular case I think because there's never going to be a 3rd state, but since it's Friday did it anyway.